ejb/WFLY-8449_Return_context_data_to_client.adoc

```diff
@@ -0,0 +1,81 @@
+= Return Context Data to EJB Clients
+:author:            Stuart Douglas
+:email:             sdouglas@redhat.com
+:toc:               left
+:icons:             font
+:keywords:          comma,separated,tags
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+This will allow for some data that has been placed in the InvocationContext by an EJB or Interceptor to
+be returned to the client so it can be processed by client interceptors. Only data that has explicitly
+been requested by the interceptor will be returned.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-8449[WFLY-8449]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-976[EAP7-976]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+=== Affected Projects or Components
+
+ * Wildfly
+ * EJB Client
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+An `EJBClientInterceptor` will be able to request data that it is interested in by calling
+`org.jboss.ejb.client.EJBClientInvocationContext#addReturnedContextDataKey(String key)`. This tells the server that the
+client interceptor would like any data that is stored under the provided key in the context data map to be returned
+to the client.
+
+Internally this will create/add to a `Set` in the clients context data stored under
+`org.jboss.ejb.client.invocation.returned-context-data`, which will then be sent to the server as part of the invocations
+context data, so no protocol change is required.
+
+If the server receives this set of keys then after the invocation has completed it will iterate the set of keys,
+and if a corresponding key name is found in the context data of the invocation then it will be copied into the
+context data map that is returned to the remote client. Any EJBClientInterceptor will be able to examine this data
+in the invocations context data map.
+
+This approach allows for a very fine grained approach to returning data to the client, so that only data that is actually
+required is returned.
+
+=== Nice-to-Have Requirements
+
+=== Non-Requirements
+
+It is not a requirement that all data in the context data map be returned to the client. This approach would
+be problematic for several reasons:
+
+ *  Most invocations simply do not want or need this data, so for the majority of clients this would increase network
+ and CPU usage for no reason
+ * There is no requirement that data that is placed in the context data map be Serializable. The only way to actually
+ determine this is to try and serialize it. As such any working implementation of this would need to catch and ignore
+ exceptions which happen during the Serialization process, or break backwards compatibility by causing invocations that
+ previously would have worked to fail.
```

Comment:
The definition of compatibility is that code which meets the specification (whatever that may be) continues to work, and that the specification is backwards-compatible.  It is possible to write code which works yet can reasonably break because it does not follow specification.  If there is a case where the specification is wrong, then we would retroactively update the specification and fix any interim problems as bugs.

Therefore we're not breaking compatibility because previously working invocations begin to fail.  If anything it would be because the specification did not evolve compatibly.  In this case we broke compatibility when we stopped sending context data back to the client in JBoss AS 7.  So, only selectively sending back context data is already breaking compatibility with our previous specification (such as it is).

That said, evolving the specification to not require serializable data in every key _is_ a backwards-compatible change (as long as we accept the incompatible change of restricting what context data is sent back).  But it is not going to be possible to just catch exceptions during serialization and then carry on like nothing has happened.  I think that it must be specified that keys listed in the returned-data set _must_ be serializable, otherwise invocation may fail.  Other keys may or may not be serializable (we don't care, as they will be dropped).

Reply:
The EJB specification does not impose any serialisability requirement on context data. It is also not in any way implied it is available to remote clients. The EJB/interceptors spec uses the context data as a way to pass data between interceptors. 

Adding an additional serialisability requirement to something that is governed by the EE specs is a break in compatibility.

It is ok for us to have a selective mechanism for them to be returned, but we can't just start serialising them all.

Reply:
When I say "the specification" I mean "the way JBoss promises this stuff will work", which is generally a superset of the Java EE (and other) specifications.

Reply:
Additional constraints, sure.  The more you promise, the larger the set of allowed behaviors.  The more you constrain, the smaller the set.  We offer features and behaviors beyond Java EE, for example, and we cannot constrain Java EE-specified behavior to a subset without breaking compliance.

Reply:
Yes, but this particular bit of the document is talking about problems that arise if you just try and serialise everything. If you attempt that then you add an additional requirement that everything that is in this map is serialisable and available on the client class path, neither of which are required by the EE spec. The serialisability one is easy enough to hack around on the server side, but AFAIK marshalling cannot recover from a failure on the client side due to a class not being available. 

In addition to this the majority (99%+) of deployments will have no client interceptors that can make use of this data anyway, so the data would simply be discarded on the client side.

The reason why this has been explicitly singled out is because so far all discussion on various JIRAs has been about making the whole thing available to every invocation. This section is about the problems that arise with this approach. 

For the majority of customers that have no interest in this feature just sending all the data back means the following:
- Extra CPU usage for serialization
- Extra network utilisation for the extra data that will be discarded
- The possibility that existing applications will stop working because a server side class is not present on the client, even though the client has no interest in that class, and the server is simply using standard EE spec functionality. 

That is why this is explicitly outlined in this section. This is a feature that will not see broad use, if you are not using it then it should have zero impact.

Reply:
The issue with backwards compatibility also has other dimension - in between minor releases of EAP, we promise the users their applications will keep on working without outsides change in behaviour inside single major release. Breaking changes should not happen. If somebody starts using their new feature in the existing application wrongly and it breaks, however, that's a different story.

So, to be concise, applications that worked in EAP 7.x have to work in 7.2 out of the box without change in behaviour.

Reply:
OK. Even with the regex solution, we'd technically we'd be breaking compatibility with any existing app which happened to use our exact key name (unlikely as that may be); it might be a good idea to document (somewhere) that undefined attachment keys starting with "jboss" or "org.jboss" are reserved and are not part of any compatibility guarantees.

Reply:
Agreed.